src/router.rs

```diff
@@ -53,19 +54,52 @@ impl RouterProxy {
                 msg_sender: msg_sender,
                 wakeup_sender: wakeup_sender,
             }),
+            shutdown: Mutex::new(false),
```

Comment:
Any time I see two locks, my deadlock alarms go off. Any chance you could make this an `AtomicBool` instead?

Reply:
Good point, and what I've done is put the boolean into the other locked data structure, what do you think(I find something like `compare_and_swap` really quite hard to work with)?